src/maintainer/knowledge_base.rst

```diff
@@ -1028,9 +1028,126 @@ In order to qualify as a noarch python package, all of the following criteria mu
+relies on three concepts:
+
+1.  `Virtual packages <https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html>`__. 
+    Prefixed with a double underscore, they are used by conda to represent system properties as
+    constraints for the solver at install-time. We will use ``__linux``, ``__win`` or ``__osx``,
+    which are only present when the running platform is Linux, Windows, or MacOS, respectively.
+    ``__unix`` is present in both Linux and MacOS. Note that this feature is **only fully available
+    on conda 4.10 or above**.
+2.  ``conda-forge.ymls``'s :ref:`noarch_platforms` option.
+
+The idea is to generate different noarch packages for each OS needing different dependencies.
+Let's say you have a pure Python package, perfectly eligible for ``noarch: python``, but on Windows
+it requires ``windows-only-dependency``. You might have something like:
+
+.. code-block:: yaml
+  :caption: recipe/meta.yaml (original)
+
+  name: package
+  source:
+    # ...
+  build:
+    number: 0
+  requirements:
+    # ...
+    run:
+      - python
+      - numpy
+      - windows-only-dependency  # [win]
+
+Being non-noarch, this means that the build matrix will include at least 12 outputs: three platforms,
+times four Python versions. This gets worse with arm64, aarch64 and ppc64le in the mix. We can get it down
+to two outputs if replace it with this other approach!
+
+.. code-block:: yaml+jinja
+  :caption: recipe/meta.yaml (modified)
+
+  name: package
+  source:
+    # ...
+  build:
+    number: 0
+    # You NEED to include the platform in the build string to avoid hash collisions
+    string: "unix_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"  # [unix]
+    string: "win_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"   # [win]
+    noarch: python
+  requirements:
+    # ...
+    run:
+      - python
+      - numpy
+      - __unix  # [unix]
+      - __win   # [win]
+      - windows-only-dependency  # [win]
+
+Do not forget to specify the platform virtual packages with their selectors!
+Otherwise, the solver will not be able to choose the variants correctly.
+
+By default, conda-forge will only build ``noarch`` packages on a ``linux_64`` CI runner, so
+only the ``# [unix]`` selectors would be true. However, we can change this behaviour using
+the ``noarch_platforms`` option in ``conda-forge.yml``:
+
+.. code-block:: yaml
+  :caption: conda-forge.yml
+
+  noarch_platforms:
+    - linux_64
+    - win_64
+
+This will provide two runners per package! Perfect! All these changes require a
+feedstock rerender to be applied. See :ref:`dev_update_rerender`.
+
+If you need conditional dependencies on all three operating systems, this is how you do it:
+
+.. code-block:: yaml+jinja
+  :caption: recipe/meta.yaml
+
+  name: package
+  source:
+    # ...
+  build:
+    number: 0
+    # You NEED to include the platform in the build string to avoid hash collisions
+    string: "{{ SUBDIR.split('-')[0] }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Comment:
I'm getting a linter error with this: https://github.com/conda-forge/poetry-feedstock/pull/80#issuecomment-1354686510
Commit: https://github.com/conda-forge/poetry-feedstock/pull/80/commits/e2abf93666388971b43d25581a84b3356a2c8d41

Reply:
Hm, we might need to run `conda smithy recipe-lint` locally to see the actual error.

Reply:
```python
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/mambaforge/base/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 670, in main
    args.subcommand_func(args)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 514, in __call__
    lints, hints = lint_recipe.main(
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/lint_recipe.py", line 978, in main
    content = render_meta_yaml("".join(fh))
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/utils.py", line 90, in render_meta_yaml
    content = env.from_string(text).render(context)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 16, in top-level template code
TypeError: 'str' object is not callable
```
This is the full error.

Reply:
Ah, I think we need to split `SUBDIR` in a different `set` expression. 
See https://stackoverflow.com/questions/30515456/split-string-into-list-in-jinja

So we should have:

```yaml
{% set os_prefix, arch_suffix = SUBDIR.split('-', 1) %}
string: "{{ os_prefix }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Reply:
Hmm, still getting the same error:

```python
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/mambaforge/base/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 670, in main
    args.subcommand_func(args)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 514, in __call__
    lints, hints = lint_recipe.main(
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/lint_recipe.py", line 978, in main
    content = render_meta_yaml("".join(fh))
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/utils.py", line 90, in render_meta_yaml
    content = env.from_string(text).render(context)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 3, in top-level template code
TypeError: 'str' object is not callable
```

Recipe:

```yaml
{% set name = "poetry" %}
{% set version = "1.3.1" %}
{% set os_prefix, arch_suffix = SUBDIR.split('-', 1) %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/poetry-{{ version }}.tar.gz
  sha256: fde98462ad5dc9879393157da93092206a3411117e25a4761a41c6d08f31aea8

build:
  noarch: python
  number: 1
  # You NEED to include the platform in the build string to avoid hash collisions
  string: "{{ os_prefix }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
  script: {{ PYTHON }} -m pip install . -vv
  entry_points:
    - poetry = poetry.console.application:main
```

Reply:
Ugh, turns out `SUBDIR` is not defined. But `target_platform` works:

```yaml
  string: "{{ target_platform.split('-')[0] }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Still, a bit ugly...

Reply:
Changed the PR so it's explicit about the platforms and doesn't use that much Jinja in the string.